### PR TITLE
feat: support storing metadata for plugins

### PR DESCRIPTION
**Pull request by nic-chen**
### What this PR does / why we need it:
 support storing metadata for plugins

fix https://github.com/apache/apisix/issues/2263

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/admin/init.lua [hunk 1]
```diff
@@ -108,6 +109,11 @@ local function run()
         seg_sub_path = core.table.concat(uri_segs, "/", 7)
     end
 
+    -- plugin metadata
```
Comment (membphis) [inline apisix/admin/init.lua:112]: I do not like this process, I accept `plugin_metadata` as key.

File: apisix/admin/plugin_metadata.lua [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local pcall   = pcall
+local require = require
+local core    = require("apisix.core")
+
+local _M = {
+    version = 0.1,
```
Comment (membphis) [inline apisix/admin/plugin_metadata.lua:22]: this version is useless, let us drop it

File: apisix/schema_def.lua [hunk 3]
```diff
@@ -602,6 +602,11 @@ _M.stream_route = {
     }
 }
 
+_M.plugin_metadata = {
```
Comment (membphis) [inline apisix/schema_def.lua:605]: I think we can move this to `plugin`.

Here is an example:

```
local _M = {
    version = 0.1,
    priority = 100,
    name = plugin_name,
    schema = schema,
    schema_metadata = {
        -- ... ...
    }
}
```

Comment (membphis) [reply apisix/schema_def.lua:599]: do we need to delete those code? it should be wrong.

File: doc/admin-api.md [hunk 4]
```diff
@@ -660,4 +661,33 @@ Config Example:
...
+|DELETE   |/apisix/admin/plugin_metadata/{plugin_name}|NULL|Remove resource|
+
+> Request Body Parameters：
+
+An object with at least one property and at most 10 properties.
+
+
+Config Example:
+
+```shell
```
Comment (membphis) [inline doc/admin-api.md:685]: we need a real example, how about we use plugin `example-plugin`

File: doc/admin-api.md [hunk 5]
```diff
@@ -660,4 +661,36 @@ Config Example:
...
+
+|Method      |Request URI|Request Body|Description        |
+|---------|-------------------------|--|------|
+|GET      |/apisix/admin/plugin_metadata/{plugin_name}|NULL|Fetch resource|
+|PUT      |/apisix/admin/plugin_metadata/{plugin_name}|{...}|Create resource by plugin name|
+|DELETE   |/apisix/admin/plugin_metadata/{plugin_name}|NULL|Remove resource|
+
+> Request Body Parameters：
+
+An object with at least one property and at most 10 properties.
```
Comment (membphis) [inline doc/admin-api.md:680]: wrong comments, please update it

Comment (membphis) [reply doc/admin-api.md:680]: And we need to add some doc in `how to develop plugin`, the developer need it:

https://github.com/apache/apisix/blob/4f83f09eebbf10e5b33765c6b331e42fdc3862e3/doc/plugin-develop.md

**Review by membphis — COMMENTED** [review]
BTW, we can update the plugin `log-rotate` in this way. use a new PR to do this.

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@gxthrj please take a look at this PR

**Review by gxthrj — APPROVED** [review]

**Pull request merged**